### DISCUSSION TITLE
How do I build JReleaser from source?

### DISCUSSION THREAD

**Discussion by ge0ffrey** [jreleaser/jreleaser#1352] [Q&A] [answered]
How do I build JReleaser from source?

Things I 've tried:

- `./gradlew build` with JDK 17 (Temurin): Hangs for at least 2 hours on `Generating wrapper`. Even with deleting `.gradle` and `~/.gradle`.

```
$ ./gradlew build
⬇️  Downloading https://services.gradle.org/distributions/gradle-6.9.4-bin.zip
📦 Unpacking gradle-6.9.4-bin.zip
📝 Creating wrapper project
👷‍ Generating wrapper
// Hang forever
```

- `./gradlew build` with JDK 11 (Temurin)

```
Error: LinkageError occurred while loading main class GradleWrapperDownloader
        java.lang.UnsupportedClassVersionError: GradleWrapperDownloader has been compiled by a more recent version of the Java Runtime (class file version 61.0), this version of the Java Runtime only recognizes class file versions up to 55.0
Error: Could not find or load main class org.gradle.wrapper.GradleWrapperMain
```

- `gradle build` with gradle 8.1.1 with JDK 17: enforcer fails on JDK version

- `gradle build` with gradle 8.1.1 with JDK 11: fails with

```
Failed to apply plugin 'org.kordamp.gradle.checkstyle'.
> Could not create task ':aggregateJacocoReport'.
   > 'void org.gradle.api.reporting.ConfigurableReport.setEnabled(boolean)'

```

-  `gradle build` with gradle 6.9.4 with JDK 17: fails with

```
> startup failed:
  General error during semantic analysis: Unsupported class file major version 61
  
  java.lang.IllegalArgumentException: Unsupported class file major version 61

```

OS: Linux

**Comment by ge0ffrey** [comment, marked as answer]
Now trying `gradle build` with gradle 6.9.4 with JDK 11. That seems to work.

**Reply by ge0ffrey** [reply]
BUILD SUCCESSFUL in 5m 56s

**Reply by aalmiray** [reply]
Build instructions are found here https://github.com/jreleaser/jreleaser/blob/main/CONTRIBUTING.adoc

**Reply by aalmiray** [reply]
I understand why the linkage error happened. The fix is to delete the `.class` file found inside the `gradle` directory and try again. 

There might be a way to upgrade the build to Gradle 7 or 8 so that Java 17 may be used. Can't be done easily because there are build dependencies that if upgraded would cause binary incompatibilities, prompting for a bump to major version 2. We're not there yet.

**Reply by ge0ffrey** [reply]
Rome wasn't build in a day :) I 've provided a PR to link the build instructions from the README, where I was looking for them :)

PR: https://github.com/jreleaser/jreleaser/pull/1353

**Reply by ge0ffrey** [reply]
FWIW, the instructions in that contribution.adoc didn't work for me (see line gradlew + JDK 11 above), but there's a ton of interesting information in there that would have (and still will) saved me a bunch of time.

**Reply by aalmiray** [reply]
Right. It didn't work because the first try created a class with Java 17 bytecode. The 2nd attempt failed because Java 11 could not load the class. 

The reasons for Java 12 and Gradle 6.9.4 being strict requirements are a bit convoluted to explain. The expectation for building the project is that contributors would look for the CONTRIBUTING doc before writing code or attempting a build 😉

I'll explore an option to upgrade the build to Gradle 8 once `v1.6.0` comes out. Hopefully this change won't break binary compatibility nor it will make build complexity explode.

**Reply by aalmiray** [reply]
@ge0ffrey build has been updated to Gradle 7.6.1. It's now possible to use Java 17 to build. The issue with unmatched bytecode wit the Gradle wrapper bootstrapper is also fixed.
